File: backend/medicalapp/src/main/java/com/medicalapp/entity/Patient.java
package com.medicalapp.entity;

import com.medicalapp.entity.enums.Gender;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Type;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "patients")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Patient {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false, unique = true)
    private User user;

    @Column(name = "first_name", nullable = false, length = 100)
    private String firstName;

    @Column(name = "last_name", nullable = false, length = 100)
    private String lastName;

    @Column(name = "birth_date")
    private LocalDate birthDate;

    @Enumerated(EnumType.STRING)
    @Column(columnDefinition = "gender")
    private Gender gender;

    @Type(JsonType.class)
    @Column(columnDefinition = "jsonb")
    @Builder.Default
    private List<String> allergies = new ArrayList<>();

    @Type(JsonType.class)
    @Column(name = "chronic_diseases", columnDefinition = "jsonb")
    @Builder.Default
    private List<String> chronicDiseases = new ArrayList<>();

    @ManyToMany(mappedBy = "patients")
    @Builder.Default
    private List<Doctor> doctors = new ArrayList<>();

    @OneToMany(mappedBy = "patient")
    @Builder.Default
    private List<Prescription> prescriptions = new ArrayList<>();
}

File: android/app/src/main/java/com/example/myapplication/domain/Patient.kt
package com.example.myapplication.domain

import java.util.Date

data class Patient (
    val id: Int,
    val firstName: String,
    val lastName: String,
    val birthDate: Date,
    val gender: Gender,
    val allergies: List<String> = emptyList(),
    val chronicDiseases: List<String> = emptyList()
) {
    val age: Int
        get() {
            val now = Date()
            val diff = now.time - birthDate.time
            return (diff / (1000L * 60 * 60 * 24 * 365)).toInt()
        }
}

enum class Gender {
    MALE,
    FEMALE
}

